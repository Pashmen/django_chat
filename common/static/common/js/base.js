const isDebugStr = document.getElementById("is_debug").dataset.isDebug;

const emptyFunc = function() {};
if (isDebugStr === "false") {
    console.log = console.info = console.error = console.debug = emptyFunc;
}

////////
async function reloadWrapper() {
    location.reload();
}

window.onerror = function() {
    const timeout = 5000;
    const message = (
        `Something happened.\n`
        + `The page will be reloaded after ${timeout/1000} sec`
    );
    const notificationNode = document.createElement("div");
    notificationNode.classList.add("notification", "error");
    notificationNode.innerHTML = message;
    document.body.prepend(notificationNode);

    setTimeout(
        reloadWrapper,
        timeout
    );
};


////////
window.addEventListener("storage", storageMessageHandler);

const logoutLink = document.getElementById("logout");
logoutLink.onclick = broadcastAboutLogout;

function broadcastAboutLogout() {
    localStorage.setItem("logout", "true");
    localStorage.removeItem("logout");
}

function storageMessageHandler(ev) {
    console.info("storageMessageHandler: ", ev);
    if (ev.key === "logout") {
        window.location.reload();
    } else {
        console.error("Invalid command");
    }
}


////////
/* istanbul ignore next */
if (navigator.userAgent.includes("jsdom")) {
    module.exports = {
        isDebugStr: isDebugStr,
        logoutLink: logoutLink,
        broadcastAboutLogout: broadcastAboutLogout,
        storageMessageHandler: storageMessageHandler,
        emptyFunc: emptyFunc,
        reloadWrapper: reloadWrapper,
    };
}
