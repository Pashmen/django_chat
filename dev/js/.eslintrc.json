{
    "env": {
        "browser": true,
        "es2021": true
    },
    "rules": {
        "radix": [
            "error",
            "always"
        ],
        "camelcase": [
            "error",
            {
                "properties": "always"
            }
        ],
        "no-multiple-empty-lines": [
            "error",
            {
                "max": 2
            }
        ],
        "no-undef": [
            "warn"
        ],
        "no-unused-vars": [
            "warn"
        ],
        "max-len": [
            "warn",
            {
                "code": 80
            }
        ],
        "indent": [
            "warn",
            4
        ],
        "semi": [
            "warn",
            "always"
        ],
        "quotes": [
            "warn",
            "double"
        ]
    },
    "extends": "eslint:recommended",
    "parserOptions": {
        "sourceType": "module"
    },
    "ignorePatterns": [
        "**/coverage/**/*.js",
        "**/node_modules/**/*.js"
    ],
    "overrides": [
        {
            "files": [
                "**/*.js"
            ],
            "env": {
                "jest": true,
                "node": true
            },
            "plugins": [
                "jest"
            ],
            "rules": {
                "jest/no-identical-title": "error",
                "jest/valid-expect": "error"
            }
        }
    ]
}
