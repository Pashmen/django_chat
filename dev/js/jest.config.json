{
  "verbose": true,
  "testEnvironment": "jsdom",
  "setupFilesAfterEnv": [
    "<rootDir>/tests/setup.js"
  ],
  "globalSetup": "<rootDir>/tests/global_setup.js",
  "testMatch": [
    "**/tests/**/test_*.js"
  ],
  "restoreMocks": true,
  "moduleDirectories": [
    "node_modules",
    "utils",
    "../../common/static",
    "../../messages/static",
    "../../accounts/static"
  ],
  "collectCoverageFrom": [
    "../../**/static/**/*.js",
    "!**/node_modules/**"
  ],
  "coverageReporters": [
    "lcov",
    "text"
  ],
  "coverageThreshold": {
    "global": {
      "branches": 30,
      "functions": 30,
      "lines": 20,
      "statements": 20
    }
  },
  "reporters": [
    "jest-progress-bar-reporter"
  ]
}
