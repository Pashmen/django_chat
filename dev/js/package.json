{
  "name": "frusale-main",
  "devDependencies": {
    "@babel/core": "7.15.8",
    "@babel/preset-env": "7.15.8",
    "babel-jest": "27.2.5",
    "eslint": "7.32.0",
    "eslint-plugin-jest": "25.2.1",
    "jest": "27.2.4",
    "jest-progress-bar-reporter": "1.0.21",
    "prettier": "2.4.1"
  }
}
