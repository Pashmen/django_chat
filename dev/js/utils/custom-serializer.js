const prettier = require("prettier");

module.exports = {
    test(value) {
        return value instanceof Element;
    },
    print(value) {
        const s = prettier.format(
            value.outerHTML, {parser: "html"}
        );
        if (s.split("\n").length === 2) {
            return s;
        } else {
            return s.trimEnd();
        }
    }
};
