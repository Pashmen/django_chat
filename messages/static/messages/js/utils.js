class CustomWebSocket extends WebSocket {
    constructor(url, connect) {
        super(url);

        this.onopen = function(e) {
            console.info("ws open", e);
        };

        this.onclose = function (e) {
            const timeout = 3000;
            console.info("ws close", e);
            setTimeout(connect, timeout);
        };

        this.onerror = function (e) {
            console.error("ws error", e);
        };
    }
}

function getWsUrl() {
    const loc = window.location;
    let wsStart = "ws://";
    if (loc.protocol === "https") {
        wsStart = "wss://";
    }

    return wsStart + loc.host + "/ws" + loc.pathname;
}

////////
export {
    CustomWebSocket,
    getWsUrl
};
