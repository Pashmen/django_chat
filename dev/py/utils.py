from unittest.mock import Mock

import pytz
from django.conf import settings
from django.contrib.auth import get_user_model
from django.contrib.auth.models import AnonymousUser
from django.contrib.auth.models import User
from django.db import models
from django.shortcuts import render
from django.test import RequestFactory
from django.test import TestCase as DjangoTestCase
from django.utils import timezone


class ProjectError(Exception):
    pass


class CustomTestCase(DjangoTestCase):
    factory = RequestFactory()
    TIME_FORMAT = settings.FS_TIME_FORMAT

    def create_user(cls):
        cls.PASSWORD = "Fwer2356scdvf"
        cls.EMAIL = "user@email.com"
        cls.user = get_user_model().objects.create(
            id=1,
            email=cls.EMAIL,
            password=cls.PASSWORD
        )

    def create_p_user(cls):
        cls.EMAIL = "user@email.com"
        cls.INFO = "Info about me"
        cls.AVATAR_PATH = "/media/avatars/standart_avatar.png"
        cls.LAST_SEEN = "2020.20.20 10:10"
        cls.user = Mock(**{
            "id": 1,
            "email": cls.EMAIL,
            "info": cls.INFO,
            "avatar.url": cls.AVATAR_PATH,
            "last_seen": cls.LAST_SEEN
        })

    def convert_str_to_time(self, time_str):
        return timezone.make_aware(
            timezone.datetime.strptime(
                time_str,
                self.TIME_FORMAT
            ),
            timezone=pytz.timezone("Etc/UTC")
        )

    def seen_in_cache_set(self, user_id, moment):
        self.redis_cache.set(
            "seen_{}".format(user_id),
            moment.strftime(self.TIME_FORMAT)
        )

    def render(
        self, template_path, data={},
        is_authorised=False, request_user_attrs=None
    ):
        request = self.factory.get("/")
        if not is_authorised:
            request.user = AnonymousUser()
        else:
            request.user = User()

        if request_user_attrs != None:
            for attr_name, attr_value in request_user_attrs.items():
                setattr(request.user, attr_name, attr_value)

        return render(
            request, template_path, data
        ).content.decode("utf-8")

    def assertTimeAlmostEqual(
        self, time1, time2, eps=60
    ):
        self.assertTrue(
            abs((time1 - time2).total_seconds()) <= eps
        )

    def assertCustomSequenceEqual(
        self, seq1, seq2
    ):
        return super().assertSequenceEqual(
            tuple(seq1), tuple(seq2)
        )

    def assertStatusCode(self, url, status_code):
        response = self.client.get(url)
        self.assertEqual(response.status_code, status_code)

    def assertCustomRedirects(
        self, url, redirect_url, status_code,
        target_status_code, request_method="get", data={}
    ):
        if request_method == "get":
            response = self.client.get(url, data)
        elif request_method == "post":
            response = self.client.post(url, data)
        else:
            raise ProjectError(
                "This request method doesn't exist or isn't supported"
            )

        self.assertRedirects(
            response, redirect_url,
            status_code, target_status_code
        )

    def assertCascadeDelete(self, model, field_name):
        self.assertEquals(
            model._meta.get_field(field_name).remote_field.on_delete,
            models.CASCADE
        )
