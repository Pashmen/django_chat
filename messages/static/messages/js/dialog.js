import {CustomWebSocket, getWsUrl} from "./utils.js";


const wsUrl = getWsUrl();
let socket;

const form = document.getElementById("form");
const textInput = document.getElementById("id_text");
const dialogHolder = document.getElementById("dialog-messages");
const m = {unreadMessagesExist: false};

if (form !== null) {
    form.onsubmit = formOnSubmitHandler;
}

function formOnSubmitHandler(ev) {
    ev.preventDefault();
    const messageText = textInput.value;

    console.info("Send new message");
    socket.send(JSON.stringify({
        "command": "get_new_message",
        "message": {
            "text": messageText
        }
    }));
    textInput.value = "";
}

function focusHandler() {
    if (m.unreadMessagesExist) {
        m.sendMarkDialogAsRead();
    }
}

window.addEventListener("focus", focusHandler);

////////
function connect() {
    socket = new CustomWebSocket(wsUrl, connect);
    socket.onmessage = socketOnMessageHandler;
}

connect();

function socketOnMessageHandler(ev) {
    console.info("ws message", ev);
    const data = JSON.parse(ev.data);

    if (!("command" in data)) {
        console.warn("invalid data");
        return;
    }

    const command = data["command"];
    console.info("command", command);
    if (command === "get_new_message") {
        const message = data["message"];
        dialogHolder.appendChild(m.createMessageNode(message));
        m.checkIntegrity(data["integrity_hash"]);

        if (message["is_unread"]) {
            if (!m.unreadMessagesExist && !document.hasFocus()) {
                m.unreadMessagesExist = true;
            } else if (document.hasFocus() && m.unreadMessagesExist) {
                m.sendMarkDialogAsRead();
            }
        }
    } else if (command === "get_messages") {
        console.info("messages:", data["messages"]);
        const messages = data["messages"];
        const messagesNodes = [];
        for (const message of messages) {
            if (message["is_unread"]) {
                m.unreadMessagesExist = true;
            }
            messagesNodes.push(m.createMessageNode(message));
        }
        dialogHolder.replaceChildren(...messagesNodes);

        if (document.hasFocus() && m.unreadMessagesExist) {
            m.sendMarkDialogAsRead();
        }
    } else if (command === "check_integrity") {
        m.checkIntegrity(data["integrity_hash"]);
    } else if (command === "go_home") {
        window.location.href = "/home/";
    } else {
        console.warn("Invalid command");
    }
}


////////
m.createMessageNode = function(message) {
    console.info("createMessageNode", message["text"]);
    const messageNode = document.createElement("pre");
    messageNode.classList.add("message");
    messageNode.dataset.hash = message["hash"];
    messageNode.dataset.time = message["time"] + "Z";

    let ownershipStr;
    if (message["user_owns_message"]) {
        ownershipStr = "(from me)";
    } else {
        ownershipStr = "(to me)";
    }

    const timeNode = document.createElement("b");
    timeNode.classList.add("time");
    messageNode.append(timeNode);
    const time = new Date(messageNode.dataset.time);
    let timeStr;
    function pad(x) {
        return ("0" + x).slice(-2);
    }
    if (time.toDateString() === (new Date()).toDateString()) {
        timeStr = time.toTimeString().slice(0, 5);
    } else {
        timeStr = `${pad(time.getDate())}.${pad(time.getMonth() + 1)}`;
    }
    timeNode.appendChild(document.createTextNode(timeStr));

    const htmlText = ` ${ownershipStr}: ${message["text"]}`;
    const textNode = document.createTextNode(htmlText);
    messageNode.appendChild(textNode);

    return messageNode;
};

m.sendMarkDialogAsRead = function() {
    console.info("Send mark_dialog_as_read");
    m.unreadMessagesExist = false;
    socket.send(JSON.stringify({
        "command": "mark_dialog_as_read"
    }));
};


m.checkIntegrity = function(serverIntegrityHash) {
    console.info("checkIntegrity");
    let integrityHash = 0;
    document.querySelectorAll(".message").forEach(
        messageNode => integrityHash += parseInt(
            messageNode.dataset.hash,
            10
        )
    );

    if (integrityHash !== serverIntegrityHash) {
        console.info("Send give_messages");
        socket.send(JSON.stringify({
            "command": "give_messages"
        }));
    }
};


////////
/* istanbul ignore next */
if (navigator.userAgent.includes("jsdom")) {
    module.exports = {
        wsUrl: wsUrl,
        form: form,
        formOnSubmitHandler: formOnSubmitHandler,
        socket: socket,
        textInput: textInput,
        connect: connect,
        focusHandler: focusHandler,
        socketOnMessageHandler: socketOnMessageHandler,
        m: m
    };
}
