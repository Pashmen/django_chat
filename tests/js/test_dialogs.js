import * as utils from "messages/js/utils.js";

const serializer = require("custom-serializer");
expect.addSnapshotSerializer(serializer);

let dialogs;
let dialogsHolder;
let dialogsLink; document.getElementById("dialogs-link");

function loadModule() {
    jest.isolateModules(() => {
        dialogs = require("messages/js/dialogs.js");
        dialogs.socket.close();
        dialogs.socket.send = jest.fn();
    });
}

function standartSetUpDialogs() {
    const dialog = {
        "id": 1, "hash": 300, "is_unread": false,
        "text": "some text"
    };

    dialogsHolder.appendChild(dialogs.m.createDialogNode(dialog));
    dialog["id"] = 2;
    dialog["hash"] = 200;
    dialog["is_unread"] = true;
    dialogsHolder.appendChild(dialogs.m.createDialogNode(dialog));
}


beforeEach(() => {
    document.body.innerHTML =
`<a href="/url/" id="dialogs-link">Dialogs</a>
<div id="dialogs"></div>`;

    loadModule();
    dialogsHolder = document.getElementById("dialogs");
    dialogsLink = document.getElementById("dialogs-link");
    console.warn = jest.fn();
});


test("wsUrl", () => {
    expect(dialogs.wsUrl).toBe(utils.getWsUrl());
});

test("connect", () => {
    expect(dialogs.socket.onmessage).toBe(dialogs.socketOnMessageHandler);
});

describe("socketOnMessageHandler", () => {
    test("invalid", () => {
        let ev = {data: JSON.stringify({})};
        dialogs.socketOnMessageHandler(ev);
        expect(console.warn.mock.calls[0][0]).toBe("invalid data");

        ev = {data: JSON.stringify({"command": "pampam"})};
        dialogs.socketOnMessageHandler(ev);
        expect(console.warn.mock.calls[1][0]).toBe("Invalid command");
    });

    describe("get_new_message", () => {
        const data = {
            "command": "get_new_message",
            "dialog": {
                "id": 2, "hash": 400, "is_unread": true,
                "text": "some text"
            }
        };

        test("dialog existed", () => {
            standartSetUpDialogs();
            jest.spyOn(dialogs.m, "checkIntegrity");
            jest.spyOn(dialogs.m, "updateDialogsLink");

            dialogs.socketOnMessageHandler({data: JSON.stringify(data)});
            expect(dialogs.m.checkIntegrity).toHaveBeenCalledTimes(1);
            expect(dialogs.m.updateDialogsLink).toHaveBeenCalledTimes(1);
            expect(document.body).toMatchSnapshot();
        });

        test("dialog didn't exist", () => {
            dialogs.socketOnMessageHandler({data: JSON.stringify(data)});
            expect(document.body).toMatchSnapshot();
        });
    });

    test("delete_dialog", () => {
        standartSetUpDialogs();
        jest.spyOn(dialogs.m, "checkIntegrity");
        jest.spyOn(dialogs.m, "updateDialogsLink");
        const data = {
            "command": "delete_dialog",
            "dialog_id": 1
        };

        dialogs.socketOnMessageHandler({data: JSON.stringify(data)});
        expect(dialogs.m.checkIntegrity).toHaveBeenCalledTimes(1);
        expect(dialogs.m.updateDialogsLink).toHaveBeenCalledTimes(1);
        expect(document.body).toMatchSnapshot();

        data["dialog_id"] = 2;
        dialogs.socketOnMessageHandler({data: JSON.stringify(data)});
        expect(document.body).toMatchSnapshot();
    });

    test("get_dialogs", () => {
        standartSetUpDialogs();
        jest.spyOn(dialogs.m, "updateDialogsLink");
        const data = {
            "command": "get_dialogs",
            "dialogs": [
                {
                    "id": 2, "hash": 500, "is_unread": true,
                    "text": "some text"
                }
            ]
        };

        dialogs.socketOnMessageHandler({data: JSON.stringify(data)});
        expect(dialogs.m.updateDialogsLink).toHaveBeenCalledTimes(1);
        expect(document.body).toMatchSnapshot();
    });

    test("mark_dialog_as_read", () => {
        standartSetUpDialogs();
        jest.spyOn(dialogs.m, "checkIntegrity");
        jest.spyOn(dialogs.m, "updateDialogsLink");
        const data = {
            "command": "mark_dialog_as_read",
            "dialog_id": 2
        };

        dialogs.socketOnMessageHandler({data: JSON.stringify(data)});
        expect(dialogs.m.updateDialogsLink).toHaveBeenCalledTimes(1);
        expect(dialogs.m.checkIntegrity).toHaveBeenCalledTimes(1);
        expect(document.body).toMatchSnapshot();
    });

    test("check_integrity", () => {
        dialogs.m.checkIntegrity = jest.fn();
        const ev = {data: JSON.stringify({
            "command": "check_integrity",
            "integrity_hash": 100
        })};

        dialogs.socketOnMessageHandler(ev);
        expect(dialogs.m.checkIntegrity.mock.calls[0][0]).toBe(100);
    });

    test("go_home", () => {
        const oldValue = window.location;
        delete window.location;
        window.location = {href: ""};
        const ev = {data: JSON.stringify({"command": "go_home"})};

        dialogs.socketOnMessageHandler(ev);
        expect(window.location.href).toBe("/home/");

        window.location = oldValue;
    });
});

test("sendDeleteDialog", () => {
    const ev = {
        preventDefault: jest.fn(),
        "srcElement": {id: 123}
    };

    dialogs.m.sendDeleteDialog(ev);
    expect(ev.preventDefault).toHaveBeenCalled();
    expect(dialogs.socket.send.mock.calls[0][0]).toEqual(
        JSON.stringify({
            "command": "delete_dialog",
            "dialog_id": 123
        })
    );
});

test("createDialogNode", () => {
    const dialog = {
        "id": 3, "hash": 321, "is_unread": false,
        "text": "some text"
    };
    let dialogNode = dialogs.m.createDialogNode(dialog);
    expect(dialogNode).toMatchSnapshot();

    dialog["is_unread"] = true;
    dialogNode = dialogs.m.createDialogNode(dialog);
    expect(dialogNode).toMatchSnapshot();
});

test("checkIntegrity", () => {
    standartSetUpDialogs();

    dialogs.m.checkIntegrity(501);
    expect(dialogs.socket.send).toHaveBeenCalledTimes(0);

    dialogs.m.checkIntegrity(500);
    expect(dialogs.socket.send.mock.calls[0][0]).toBe(
        JSON.stringify({"command": "give_dialogs"})
    );
});

describe("updateDialogsLink", () => {
    test("no unread", () => {
        dialogsLink.classList.add("unread-dialogs-exist");
        dialogs.m.updateDialogsLink();
        expect(dialogsLink).toMatchSnapshot();
    });

    test("with unread", () => {
        standartSetUpDialogs();

        dialogs.m.updateDialogsLink();
        expect(dialogsLink).toMatchSnapshot();
    });
});
