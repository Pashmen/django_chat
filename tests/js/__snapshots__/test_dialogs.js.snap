// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDialogNode 1`] = `
<form class="dialog" id="3" data-hash="321">
  <a href="/dialogs/u3">3: some text</a
  ><button type="submit" name="Delete">x</button><br />
</form>
`;

exports[`createDialogNode 2`] = `
<form class="dialog" id="3" data-hash="321">
  <a href="/dialogs/u3" class="is-unread">3: some text</a
  ><button type="submit" name="Delete">x</button><br />
</form>
`;

exports[`socketOnMessageHandler delete_dialog 1`] = `
<body>
  <a href="/url/" id="dialogs-link" class="in-dialogs unread-dialogs-exist"
    >Dialogs (1)</a
  >
  <div id="dialogs">
    <form class="dialog" id="2" data-hash="200">
      <a href="/dialogs/u2" class="is-unread">2: some text</a
      ><button type="submit" name="Delete">x</button><br />
    </form>
  </div>
</body>
`;

exports[`socketOnMessageHandler delete_dialog 2`] = `
<body>
  <a href="/url/" id="dialogs-link" class="in-dialogs">Dialogs</a>
  <div id="dialogs"></div>
</body>
`;

exports[`socketOnMessageHandler get_dialogs 1`] = `
<body>
  <a href="/url/" id="dialogs-link" class="in-dialogs unread-dialogs-exist"
    >Dialogs (1)</a
  >
  <div id="dialogs">
    <form class="dialog" id="2" data-hash="500">
      <a href="/dialogs/u2" class="is-unread">2: some text</a
      ><button type="submit" name="Delete">x</button><br />
    </form>
  </div>
</body>
`;

exports[`socketOnMessageHandler get_new_message dialog didn't exist 1`] = `
<body>
  <a href="/url/" id="dialogs-link" class="in-dialogs unread-dialogs-exist"
    >Dialogs (1)</a
  >
  <div id="dialogs">
    <form class="dialog" id="2" data-hash="400">
      <a href="/dialogs/u2" class="is-unread">2: some text</a
      ><button type="submit" name="Delete">x</button><br />
    </form>
  </div>
</body>
`;

exports[`socketOnMessageHandler get_new_message dialog existed 1`] = `
<body>
  <a href="/url/" id="dialogs-link" class="in-dialogs unread-dialogs-exist"
    >Dialogs (1)</a
  >
  <div id="dialogs">
    <form class="dialog" id="2" data-hash="400">
      <a href="/dialogs/u2" class="is-unread">2: some text</a
      ><button type="submit" name="Delete">x</button><br />
    </form>
    <form class="dialog" id="1" data-hash="300">
      <a href="/dialogs/u1">1: some text</a
      ><button type="submit" name="Delete">x</button><br />
    </form>
  </div>
</body>
`;

exports[`socketOnMessageHandler mark_dialog_as_read 1`] = `
<body>
  <a href="/url/" id="dialogs-link" class="in-dialogs">Dialogs</a>
  <div id="dialogs">
    <form class="dialog" id="1" data-hash="300">
      <a href="/dialogs/u1">1: some text</a
      ><button type="submit" name="Delete">x</button><br />
    </form>
    <form class="dialog" id="2" data-hash="200">
      <a href="/dialogs/u2" class="">2: some text</a
      ><button type="submit" name="Delete">x</button><br />
    </form>
  </div>
</body>
`;

exports[`updateDialogsLink no unread 1`] = `
<a href="/url/" id="dialogs-link" class="in-dialogs">Dialogs</a>

`;

exports[`updateDialogsLink with unread 1`] = `
<a href="/url/" id="dialogs-link" class="in-dialogs unread-dialogs-exist"
  >Dialogs (1)</a
>
`;
